inference-engine/src/auto_plugin/auto_plugin.cpp

```diff
@@ -193,6 +283,90 @@ ConfigType AutoInferencePlugin::GetSupportedConfig(const ConfigType&  config,
     return supportedConfig;
 }
 
+DeviceInformation AutoInferencePlugin::SelectDevice(const std::vector<DeviceInformation>& metaDevices) {
+    if (metaDevices.empty()) {
+        IE_THROW(NotFound) << "No available device to select in AUTO plugin";
+    }
+    if (metaDevices.size() == 1) {
+        return metaDevices.at(0);
+    }
+
+    std::vector<DeviceInformation> CPU;
+    std::vector<DeviceInformation> GPU;
+
+    for (auto& item : metaDevices) {
+        if (item.deviceName.find("CPU") == 0) {
+            CPU.push_back(item);
+            continue;
+        }
+        if (item.deviceName.find("GPU") == 0) {
+            GPU.push_back(item);
+            continue;
+        }
+    }
+
+    if (CPU.empty() && GPU.empty()) {
+        IE_THROW(NotFound) << "No available device found";
+    }
+
+    // Sort GPU by name: GPU.2 > GPU.1 > GPU.0 > GPU, so we always choose the GPU[0] as best device
+    std::sort(GPU.begin(), GPU.end(), [](const DeviceInformation& a, const DeviceInformation& b)->bool{return b.deviceName < a.deviceName;});
+
+    if (!GPU.empty()) {
+        return GPU[0];
+    }
+    if (CPU.empty()) {
+        IE_THROW() << "Cannot select any device";
+    }
+    return CPU[0];
+}
+
+DeviceInformation AutoInferencePlugin::SelectDevice(const InferenceEngine::CNNNetwork&    network,
+                                                    const std::vector<DeviceInformation>& metaDevices) {
+    if (metaDevices.empty()) {
+        IE_THROW(NotFound) << "No available device to select in AUTO plugin";
+    }
+    if (metaDevices.size() == 1) {
+        return metaDevices.at(0);
+    }
+
+    auto networkPrecision = GetNetworkPrecision(network);
+
+    std::vector<DeviceInformation> CPU;
+    std::vector<DeviceInformation> GPU;
+
+    for (auto& item : metaDevices) {
+        if (item.deviceName.find("CPU") == 0) {
+            CPU.push_back(item);
+            continue;
+        }
+        if (item.deviceName.find("GPU") == 0) {
+            GPU.push_back(item);
+            continue;
+        }
+    }
+
+    if (CPU.empty() && GPU.empty()) {
+        IE_THROW(NotFound) << "No available device found";
+    }
+
+    // Sort GPU by name: GPU.2 > GPU.1 > GPU.0 > GPU, so we always choose the GPU[0] as best device
+    std::sort(GPU.begin(), GPU.end(), [](const DeviceInformation& a, const DeviceInformation& b)->bool{return b.deviceName < a.deviceName;});
+
+    for (auto&& item : GPU) {
+        std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
+        auto res = std::find(capability.begin(), capability.end(), networkPrecision);
```

Comment:
having no e.g. int8 in the capabilities doesn't imply the device is not supporting the int8 execution.
but yes, it can be less efficient, than with int8 (HW) support.

Reply:
so is it ok? or add comment to code?